레드박스 기부, 개인 기부

Bin-Choi opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
1. 레드박스 기부 기능 구현
2. 개인에게 기부 기능 구현

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
Config 를 활용한 donation 타입 별 서비슬 분기 처리에서 좀 더 좋은 방법이 있을까요?
table을 통합하면서 더 좋은 방법이 있을 것 같아요

Bin-Choi committed:
feat: 레드박스 (RB-182)
- Entity, Controller, Service, Repository 생성

Bin-Choi committed:
feat: 레드박스 엔티티 (RB-189)
- 헌혈증 수 관리를 위한 원시값 포장
- Count 클래스 생성
- 헌혈증 갯수 증감 기능

Bin-Choi committed:
feat: 레드박스 정보 조회
- Controller, Service, DTO 생성
- 레드박스 누적 count 조회

Bin-Choi committed:
feat: 레드박스 기부 (RB-188)
- redbox 기부 내역, 기부 상세 dto 작성

Bin-Choi committed:
Merge pull request #22 from prgrms-be-devcourse/feature/redbox

Feature/redbox

Bin-Choi committed:
fix: 레드박스
- redbox 기부시 count 변경 로직 수정
- count 타입 변경
- user 클래스 import 제거

Bin-Choi committed:
feat: 레드박스 기부
- 개인 기부, 게시글 기부, 레드박스 기부를 하나의 엔드포인트로 처리하는 방식으로 설계
- Config 파일 생성
- 레드박스 기부 구조 생성

jhyngu committed:
fix: Redbox 엔티티 매핑 오류 수정

- Redbox 엔티티에서 중복된 컬럼 매핑 문제 해결
- totalCount와 currentCount 필드에 @AttributeOverride 적용
  - total_count와 current_count로 컬럼 이름 지정

jhyngu committed:
refactor: RedboxService 명명 수정 및 불필요한 DTO 삭제

- RedboxSerivce -> RedboxService로 클래스명 수정
- DonationServiceConfig에서 RedboxService 반영
- 사용하지 않는 DTO(RedboxInfoResponse) 삭제

jhyngu committed:
feat: 레드박스 누적 개수 조회 API 추가 (RB-233)

- RedboxController에 `/redbox/total-count` 엔드포인트 추가
- RedboxService에 누적 개수 조회 로직 추가
- TotalCountResponse DTO 추가

jhyngu committed:
refactor: Redbox 엔티티 컬럼명 변경

- `total_count` -> `total_count_value`, `current_count` -> `current_count_value`로 컬럼명 변경

jhyngu committed:
feat: Redbox 관련 예외 추가
- REDBOX_NOT_FOUND

jhyngu committed:
feat: RedboxNotFoundException 커스텀 예외 클래스 추가

jhyngu committed:
feat: RedboxService에서 RedboxNotFoundException 예외 처리 추가

jhyngu committed:
feat: 도움을 받은 환자 수 조회 엔드포인트 추가
- /redbox/received-patients

jhyngu committed:
feat: RedboxReceiptGroup 엔티티 생성

jhyngu committed:
feat: RedboxReceiptGroupRepository에 받은 환자 수 조회를 위한 count 메서드 추가

jhyngu committed:
feat: 받은 환자 수 조회 로직 추가
- getReceivedPatientsCount: 레드박스_receipt_groups 테이블에서 받은 환자 수 조회

jhyngu committed:
Merge pull request #31 from prgrms-be-devcourse/feature/get-dashboard-info

레드박스 헌혈증 누적 개수 조회, 기부 받은 환자 수 조회 API 구현

Bin-Choi committed:
feat: 레드박스 기부(RB-181)
- 레드박스 기부 시 기부자 수량 검증
- 레드박스 기부 시 헌혈증 소유자 변경

Bin-Choi committed:
test: 레드박스 기부(RB-181)
- 테스트 코드 작성

Bin-Choi committed:
feat: 레드박스 기부(RB-181)
- 추상클래스를 사용하여 보유 레드카드 반환, 검증 로직 공통으로 처리
- 기부시 기부이력 생성 로직 추가

Bin-Choi committed:
test: 레드박스 기부(RB-181)
- 책임 변경에 따른 테스트 코드 변경 & 제거

leeys9423 committed:
Merge remote-tracking branch 'origin/devBE' into feature/donate

# Conflicts:
#	backend/src/main/java/com/redbox/global/exception/ErrorCode.java
#	backend/src/test/java/com/redbox/domain/redcard/entity/RedCardTest.java

leeys9423 committed:
merge: devBE

leeys9423 committed:
merge: devBE

Bin-Choi committed:
test: 레드박스 기부(RB-181)
- repository Test 생성

Bin-Choi committed:
Merge remote-tracking branch 'origin/feature/donate' into feature/donate

leeys9423 committed:
style: entity 이름 변경

Bin-Choi committed:
refactor: 레드박스 기부(RB-181)
- 기부 테이블 통일

Bin-Choi committed:
refactor: 레드박스 기부(RB-181)
- 레드박스의 기부 횟수 조회 변경

Bin-Choi committed:
refactor: 레드박스 기부(RB-181)
- 기부 내역 생성 모듈화

Bin-Choi committed:
feat: 기부 테이블 타입 추가
- DonationType 생성

Bin-Choi committed:
feat: 유저에게 기부하기

Bin-Choi commented on backend/src/main/java/com/redbox/domain/donation/application/AbstractDonationService.java:
```diff
@@ -0,0 +1,75 @@
+            donationDetailRepository.save(donationDetail);
+        }
+    }
+
```
게시글 기부 기능 완성하면서 ErrorCode 처리 해두겠습니다

Bin-Choi replied:
완료

jhyngu commented:
👍

leeys9423 commented on backend/src/main/java/com/redbox/domain/donation/entity/DonationGroup.java:
```diff
@@ -0,0 +1,38 @@
+    private Long id;
+    private Long donorId;
+    private Long receiverId;
+    private DonationType donationType;
```
@Enumerated(EnumType.STRING)
빠진 것 같아여

Bin-Choi replied:
체크감샇반디아

leeys9423 commented on backend/src/main/java/com/redbox/domain/redbox/dto/Count.java:
이 클래스를 보고 제가 궁금했던 점과 찾아 본 결과를 공유드립니다
질문)
Count에 대한 불변성은 보장해주지만, 연산이 일어날때마다 새로운 인스턴스를 생성해주는 건 메모리에 어떤 영향을 줄까요??

답) 
이런 Immutable 객체 설계에 대해 메모리 관점에서 분석해보겠습니다.
1. **장점**:
   - 스레드 안전성: 객체가 불변이므로 여러 스레드에서 안전하게 공유될 수 있습니다
   - 부수효과 방지: 상태 변경이 없어 버그 발생 가능성이 줄어듭니다

2. **단점 (메모리 관점)**:
   - 새로운 객체 생성 비용: 연산마다 새로운 객체가 힙 메모리에 할당됩니다
   - GC 부하: 사용이 끝난 이전 객체들은 가비지 컬렉션의 대상이 됩니다

3. **실제 영향**:
   - 작은 객체(이 경우 int 하나)의 생성은 상대적으로 비용이 적습니다
   - 모던 JVM의 경우:
     - Escape Analysis를 통해 필요한 경우 스택 할당으로 최적화할 수 있습니다
     - Young Generation에서의 GC는 매우 효율적입니다

4. **대안 검토**:
   ```java
   // 기존 방식 (Immutable)
   public Count add(int amount) {
       return new Count(value + amount);
   }
   
   // 대안 방식 (Mutable)
   public void add(int amount) {
       this.value += amount;
   }
   ```

**결론**: 
- 이 정도 크기의 객체에서는 메모리 영향이 미미합니다
- 불변성이 주는 장점이 약간의 메모리 오버헤드보다 더 가치있을 수 있습니다
- 다만 대량의 연산이 필요한 경우(예: 수백만 번의 연산)에는 성능 테스트를 통한 검증이 필요할 수 있습니다

해당 코드가 실제로 성능 병목이 되는지 확인하고 싶다면, JMH 같은 벤치마킹 도구로 테스트해보는 것을 추천드립니다.

Bin-Choi replied:
비슷한 부분에서 고민을 좀 했었는데요,
Count 객체가 작아서 그냥 불변객체로 때렸습니다 ㅋㅋ...

leeys9423 commented:
고생하셨습니다

jhyngu committed:
feat: 대시보드에서 "내 기부 통계" 기능 구현 (RB-166)

- `DonationController`에 `/users/my-donation-stats` 엔드포인트 추가하여 사용자 기부 통계 조회 기능 구현.
- `DonationStatsService`를 통해 총 기부한 헌혈증 개수와 도움받은 환자 수 계산 로직 추가.
- `DonationGroupRepository`에 기부 통계 조회를 위한 커스텀 쿼리 추가:
  - `sumDonationAmountByDonorId`: 총 기부 개수 계산.
  - `countDistinctReceiverIdByDonorIdAndReceiverIdNot`: 도움받은 환자 수 계산.
- 응답 구조를 위한 `MyDonationStatsResponse` DTO 생성.
- `SecurityConfig`에서 해당 엔드포인트에 `authenticated()` 설정으로 인증된 사용자만 접근 가능하도록 수정.

jhyngu committed:
Merge remote-tracking branch 'origin/feature/donate' into feature/donate

leeys9423 commented on backend/src/main/java/com/redbox/domain/donation/application/AbstractDonationService.java:
```diff
@@ -0,0 +1,75 @@
+
+    protected List<Redcard> getUsersRedCardList(DonationRequest donationRequest) {
+        Long donateUserId = userService.getCurrentUserId();
+        List<Redcard> redcardList = redcardRepository.findByUserId(donateUserId);
```
별건 아니지만, db에서 처음부터 갯수만큼 가져오는 방향은 어떻게 생각하세요?! 아래와 같은 효과가 있다고 하네요
1. 데이터베이스에서 필요한 수량만큼만 조회하므로 네트워크 트래픽이 감소
2. JVM 메모리 사용량이 줄어듦
3. 불필요한 subList 연산이 제거

Bin-Choi replied:
원래 그방법을 생각했었는데 기억이 안나는 모종의 이유로 인해서 (재활용 하려고 그랬던것 같은데...)
이렇게 왔습니다.
이유가 마땅히 생각이 안나면 변경하겠습니다~

Bin-Choi replied:
기억났습니다!!!
getUserRedCardList가 다른 곳에서도 쓰일것 같아서 (내 헌혈증 목록과 같은)
저렇게 뺐는데 validation을 저기에 넣어서 이도저도 아니게 됐네요.
안그래도 validation 시점에서 찝찝함을 느끼면서 넘어갔던 기억이 났습니다...

leeys9423 replied:
이부분에 대해서 항상 고민하는 부분인데,
공통으로 쓰는부분에 대해서는 최대한 공통으로 처리하는게 좋은건지,
비슷한기능이지만 그 기능만을 위한 기능을 따로 만들어서 사용하는게 좋은건지 모르겠어요.....

Bin-Choi replied:
추후에 메서드를 좀 더 작은 단위로 분리한 뒤 getUsersRedCardList 메서드가 재활용이 얼마나 되는지 고려해보겠습니다!!

Bin-Choi committed:
feat: 기부관련 에러코드
- 에러코드 생성
- exceptionHandler 처리

Bin-Choi committed:
Merge remote-tracking branch 'origin/feature/donate' into feature/donate

leeys9423 commented on backend/src/main/java/com/redbox/domain/donation/application/AbstractDonationService.java:
```diff
@@ -0,0 +1,75 @@
+
+@RequiredArgsConstructor
+public abstract class AbstractDonationService implements DonationService {
+    protected final UserService userService;
+    protected final RedcardRepository redcardRepository;
+    protected final RedcardService redcardService;
+    protected final DonationGroupRepository donationGroupRepository;
+    protected final DonationDetailRepository donationDetailRepository;
```
캡슐화 측면에서는 더 나은방법이지만 불필요하게 긴 private 필드 + protected getter 사용  하는 방법이 있다고 하네요~
```
public abstract class AbstractDonationService {
    private final UserService userService;
    private final RedcardRepository redcardRepository;
    // ...
    
    protected UserService getUserService() {
        return userService;
    }
    protected RedcardRepository getRedcardRepository() {
        return redcardRepository;
    }
}
```

Bin-Choi replied:
오호 주입을 protected getter로 하는건가요 그럼??

leeys9423 replied:
```
public abstract class AbstractDonationService implements DonationService {
    // private final로 선언
    private final UserService userService;
    private final RedcardRepository redcardRepository;
    private final RedcardService redcardService;
    private final DonationGroupRepository donationGroupRepository;
    private final DonationDetailRepository donationDetailRepository;

    // 생성자 주입
    protected AbstractDonationService(UserService userService, ...) {
        this.userService = userService;
        ...
    }

    // protected getter 제공
    protected UserService getUserService() {
        return userService;
    }
    protected RedcardRepository getRedcardRepository() {
        return redcardRepository;
    }
    // ...
}
```

요런 느낌으로 하면 된다고 하네요!

Bin-Choi replied:
자식 클래스에서 사용하는 필드도 부모클래스에서 pricate final 로 선언하고 자식클래스에서는
getter를 사용해서 가져오는 건가요???

leeys9423 replied:
넵맞아여

Bin-Choi committed:
refactor: 기부 리뷰반영
- enum 필드 @Enumerated 추가
- getUsersRedCardList 메서드에서 valid 분리
